lib/pf/domain/ntlm_cache.pm

```diff
@@ -0,0 +1,334 @@
+
+    my ($valid_users_file, $err) = generate_valid_users_file($domain);
+    unless($valid_users_file) {
+        my $msg = "Cannot generate valid users file ($err)";
+        $logger->error($msg);
+        return ($FALSE, $msg);
+    }
+
+    my ($ntds_file, $msg) = secretsdump($domain, $source, "-usersfile $valid_users_file");
+    return ($FALSE, $msg) unless($ntds_file);
+
+    $logger->info("Generated NTDS file $ntds_file");
+    return ($ntds_file);
+}
+
+=head2 get_sync_samaccountname
+
+Get the sAMAccountName for use in the sync based on the auth source
+
+=cut
+
+sub get_sync_samaccountname {
+    my ($domain, $source) = @_;
+    my $logger = get_logger;
+    my $config = $ConfigDomain{$domain};
+    
+    my ($connection, $LDAPServer, $LDAPServerPort ) = $source->_connect();
+
+    if (!defined($connection)) {
+        return ($FALSE, "Error communicating with the LDAP server");
+    }
+
+    # We need to fetch the sAMAccountName of the DN in the AD source
+    my $result = $connection->search(
+        base => $source->{binddn}, 
+        filter => '(sAMAccountName=*)', 
+        attrs => ['sAMAccountName'],
+    );
+
+    return ($FALSE, "Cannot find sAMAccountName of user ".$source->{binddn}) unless($result->count > 0);
+
+    my $sAMAccountName = $result->entry(0)->get_value('sAMAccountName');
+
+    return $sAMAccountName;
+}
+
+=head2 secretsdump
+
+Call the secretsdump binary and return the NTDS filename
+
+=cut
+
+sub secretsdump {
+    my ($domain, $source, $opts) = @_;
+    $opts //= "";
+    my $logger = get_logger;
+    my $config = $ConfigDomain{$domain};
+
+    my $tmpfile = File::Temp->new()->filename;
+    my $ntds_file = $tmpfile.".ntds";
+    
+    my ($sAMAccountName, $msg) = get_sync_samaccountname($domain, $source);
+    return ($FALSE, $msg) unless($sAMAccountName);
+
+    my $result;
+    eval {
+        my $command = "/usr/local/pf/addons/AD/secretsdump.py '".pf::domain::escape_bind_user_string($sAMAccountName)."':'".pf::domain::escape_bind_user_string($source->{password})."'@".$source->{host}." -just-dc-ntlm -output $tmpfile $opts";
+        $logger->debug("Executing sync command: $command");
+        $result = pf_run($command, accepted_exit_status => [ 0 ]);
+    };
+    if (!defined($result) || $@) {
+        return ($FALSE, "Can't generate hash list via secretsdump.py. Check logs for details.");
+    }
+    elsif($result =~ /Something wen't wrong/) {
+        return ($FALSE, "Cannot synchronize users hashes. Command output: $result");
+    }
+
+    return $ntds_file;
+}
+
+=head2 cache_user
+
+Populate the NTLM cache for a single user
+
+=cut
+
+sub cache_user {
+    my ($domain, $username) = @_;
+    my $logger = get_logger;
+    my $config = $ConfigDomain{$domain};
+    my $source = getAuthenticationSource($config->{ntlm_cache_source});
+    return ($FALSE, "Invalid LDAP source $config->{ntlm_cache_source}") unless(defined($source));
+    
+    if($username =~ /^host\//) {
```

Comment:
Can we do something more generic ? Like the user attribute can be mail, cn ... so can we have something like userAttributeToSamAccountName instead ?